parlai/mturk/tasks/acute_eval/run.py

```diff
@@ -112,341 +102,379 @@ def add_args(from_argv=False):
+        # b/c we released enough hits to guarantee reaching the requested num on average
+        tasks_still_needed = tasks_per_hit - len(task_data)
+        tasks_remaining = [
+            t_id
+            for t_id in range(len(self.desired_tasks))
+            if t_id not in worker_data['tasks_completed']
+        ]
+        # get any pairings with conversations this worker has not seen to fill this hit
+        tasks_chosen = [
+            t
+            for t in tasks_remaining
+            if all(
+                d_id not in worker_data['conversations_seen']
+                for d_id in self._get_dialogue_ids(self.desired_tasks[t])
             )
-            while not world.episode_done():
-                world.parley()
-
-            world.shutdown()
-
-            save_data = world.prep_save_data(workers)
-
-            if not world.did_complete():
-                return_task_data(workers[0].worker_id, task_data)
-            elif opt['block_on_onboarding_fail']:
-                # check whether workers failed onboarding
-                check_and_update_worker_approval(
-                    mturk_manager,
-                    workers[0].worker_id,
-                    opt['onboarding_threshold'],
-                    save_data,
-                )
-            return save_data
+        ]
+        if tasks_still_needed < len(tasks_chosen):
+            tasks_chosen = random.sample(tasks_chosen, tasks_still_needed)
+        worker_data['tasks_completed'].extend(tasks_chosen)
+        for t in tasks_chosen:
+            worker_data['conversations_seen'].extend(
+                self._get_dialogue_ids(self.desired_tasks[t])
+            )
+            task_data.extend(self.desired_tasks[t])
+        return task_data
 
-        # Soft-block all chosen workers
-        if not opt['is_sandbox'] and opt['softblock_list_path'] is not None:
+    def requeue_task_data(self, worker_id: int, task_data: List[Dict[str, Any]]):
+        """
+        Return task to task_queue.
+
+        If the task is an onboarding task, indicate that the worker has
+        another onboarding task to do.
+
+        :param worker_id:
+            worker id of worker who is returning task
+
+        :param task_data:
+            list of unfinished tasks to return to the queue.
+        """
+        worker_data = self._get_worker_data(worker_id)
+        for subtask_data in task_data:
+            if subtask_data['task_specs'].get('is_onboarding', False):
+                worker_data['onboarding_todo'].append(subtask_data['pair_id'])
+            else:
+                self.task_queue.put(subtask_data)
+                try:
+                    worker_data['tasks_completed'].remove(subtask_data['pair_id'])
+                    for d_id in self._get_dialogue_ids(subtask_data):
+                        worker_data['conversations_seen'].remove(d_id)
+                except ValueError:
```

Comment:
why does this happen?

Reply:
I am actually not sure why this would happen. presumably there is an issue where the worker's `tasks_completed` and `conversations_seen` are not in sync with the actual task data given to the worker. perhaps @hadasah might have more insight into why this error happened in the past?

Reply:
This entire file is technically not thread-safe. It's theoretically possible (although incredibly unlikely) that you could have a race condition which gave the same worker the same conversation twice and then have this situation. I actually think we should get rid of this though, given that it's not a product. If that were to actually happen and a turker complains, we'll just pay them.

Reply:
Thanks for the insight @hadasah, for now I'll just add a comment indicating why this might cause an exception